### ISSUE TITLE
Exception after training on Docker image e88040337fa3

### ISSUE THREAD

**Issue by sstainba** [ultralytics/yolov5#8764]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training

### Bug

After training two different custom models, both runs end with several exceptions thrown:

```
wandb: Synced 5 W&B file(s), 111 media file(s), 1 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20220728_115228-25fqw5ps/logs
Exception ignored in: <function StorageWeakRef.__del__ at 0x7fc14e083040>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/multiprocessing/reductions.py", line 38, in __del__
  File "/opt/conda/lib/python3.8/site-packages/torch/storage.py", line 636, in _free_weak_ref
AttributeError: 'NoneType' object has no attribute '_free_weak_ref'
```

This is exception is repeated about 20 times.

### Environment

Docker image e88040337fa3
Windows 11 Host using WSL2
Nvidia 3060 12GB GPU

### Minimal Reproducible Example

Doesn't appear to be specific to my work.  This happened with two different data sets.

Dataset 1:  300 images 1280 x1024,  1 Tag/Class
Command:  `train.py --rect --imgsz 1280 --img 1024 --epochs 500 --name test1 --cache --batch 4 --data /usr/src/datasets/data.yaml --weights yolov5m6.pt`

Dataset 2: 200 images 416 x 416,  1 Tag/Class
Command: `train.py --img 416 --epochs 500 --name splash --cache --batch 8 --data /usr/src/datasets/splash/data.yaml --weights yolov5s.pt`

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@sstainba this is a pytorch issue that should be resolved in later versions of torch. You can safely ignore it.

**Comment by CanisLupus518** [comment]
@glenn-jocher I am curious what the pytorch issue is.

After running into this myself and digging into this a bit more, I found documentation from nVidia related to WSL2 limitations: [here](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#known-limitations-for-linux-cuda-apps). Basically, in this scenario, uses on pin_memory=True are very limited.  This seems more like an nVidia\WSL2 issue than a pytorch issue. 
This issue exists when running directly in WSL2 or running in Docker+WSL2.

For yolov5, I was able successfully train in Docker+WSL2+CUDA only after changing pin_memory to False in utils/dataloaders.py

I wonder if you might consider exposing pin_memory as an argument to train, val, detect, etc. to allow for greater flexibility.

**Comment by glenn-jocher** [comment]
@CanisLupus518 got it, thanks for the info! I'll take a look at simply setting this to False. Last time I profiled it did provide benefits on later GPUs like T4 and V100. What CUDA device are you using?

**Comment by glenn-jocher** [comment]
@CanisLupus518 could we conditionally pin memory depending no the CUDA version or device properties?
<img width="845" alt="Screenshot 2022-09-01 at 11 19 30" src="https://user-images.githubusercontent.com/26833433/187879338-6dee9458-4aa6-4a64-b44a-ea9e2cc66a91.png">

**Comment by glenn-jocher** [comment]
@CanisLupus518 I profiled 3 epochs of VOC, got .121 hours with pin_memory=True and .126 hours with pin_memory=False. Epoch times were about 1:51 True and 1:56 false, so I think we want to enable by default, but I'll add an environment variable to allow setting.

**Comment by glenn-jocher** [comment]
@CanisLupus518 good news 😃! Your original issue may now be fixed ✅ in PR #9250. This PR allows the user to disable pin_memory during training/val using a new `PIN_MEMORY` environment variable, i.e.

```bash
PIN_MEMORY=False python train.py 
```

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by glenn-jocher** [comment]
@CanisLupus518 you're welcome! 😀 Excellent choice in GPUs, the RTX 3070 Ti is a powerhouse. Let us know if you encounter any other issues or have further suggestions. Happy training with YOLOv5 and the RTX 3070 Ti 🚀!
